p2p/src/network/rpc/p2p_network_rpc_reducer.rs

```diff
@@ -1,33 +1,83 @@
+use std::sync::Arc;
+
 use binprot::BinProtRead;
-use mina_p2p_messages::rpc_kernel::{MessageHeader, QueryHeader, RpcMethod};
+use mina_p2p_messages::{
+    rpc,
+    rpc_kernel::{
+        MessageHeader, PayloadBinprotReader as _, QueryHeader, ResponseHeader, RpcMethod,
+        RpcQueryReadError, RpcResponseReadError,
+    },
+    v2,
+    versioned::Ver,
+};
+use openmina_core::{error, fuzz_maybe, fuzzed_maybe, Substate};
+use redux::Dispatcher;
 
-use crate::{Limit, P2pLimits};
+use crate::{
+    channels::rpc::{
+        BestTipWithProof, P2pChannelsRpcAction, P2pRpcRequest, P2pRpcResponse,
+        StagedLedgerAuxAndPendingCoinbases,
+    },
+    connection::outgoing::P2pConnectionOutgoingInitOpts,
+    disconnection::{P2pDisconnectionAction, P2pDisconnectionReason},
+    Data, Limit, P2pLimits, P2pNetworkState, P2pNetworkYamuxAction, PeerId,
+};
 
 use self::p2p_network_rpc_state::P2pNetworkRpcError;
 
 use super::*;
 
 impl P2pNetworkRpcState {
-    pub fn reducer(
-        &mut self,
+    pub fn reducer<State, Action>(
+        mut state_context: Substate<Action, State, P2pNetworkState>,
         action: redux::ActionWithMeta<&P2pNetworkRpcAction>,
         limits: &P2pLimits,
-    ) {
-        match action.action() {
-            P2pNetworkRpcAction::Init { incoming, .. } => {
-                self.is_incoming = *incoming;
+    ) -> Result<(), String>
+    where
+        State: crate::P2pStateTrait,
+        Action: crate::P2pActionTrait<State>,
+    {
+        let rpc_state = state_context
+            .get_substate_mut()?
+            .find_rpc_state_mut(action.action())
+            .ok_or_else(|| format!("RPC state not found for action: {:?}", action.action()))?;
```

Comment:
should this be a bug_condition? I would assume that the enabling condition should filter the action so that it doesn't get dispatched if the state doesn't exist, right?